src/hotspot/share/gc/parallel/psOldGen.cpp

```diff
@@ -171,6 +172,49 @@ HeapWord* PSOldGen::allocate(size_t word_size) {
   return res;
 }
 
+/*
+ * Divide space into blocks, processes block begins at
+ * bottom + block_index  * _iterate_block_size.
+ * NOTE:
+ * - The initial block start address may not be a valid
+ * object address, _start_array is used to correct it.
+ *
+ * - The end address is not necessary to be object address.
+ *
+ * - If there is an object that crosses blocks, it is
+ * processed by the worker that owns the block within
+ * which the object starts.
+ *
+ */
+void PSOldGen::block_iterate(ObjectClosure* cl, uint block_index) {
+  MutableSpace *space = object_space();
+  HeapWord* bottom = space->bottom();
+  HeapWord* top = space->top();
+  HeapWord* begin = bottom + block_index * _iterate_block_size;
+
+  assert((_iterate_block_size % (ObjectStartArray::block_size)) == 0,
+         "BLOCK SIZE not a multiple of start_array block");
+
+  // iterate objects in block.
+  HeapWord* end = MIN2(top, begin + _iterate_block_size);
```

Comment:
`_iterate_block_size` is a size in bytes (or at least you use it like a size in bytes when calculating the number of blocks), so before you can add it you need to convert it to a size in words like this:

```suggestion
  size_t block_word_size = _iterate_block_size / HeapWordSize;
  HeapWord* begin = bottom + block_index * block_word_size;
 
  assert((block_word_size % (ObjectStartArray::block_size)) == 0,
          "BLOCK SIZE not a multiple of start_array block");
 
  // iterate objects in block.
  HeapWord* end = MIN2(top, begin + block_word_size);
```

Reply:
Nice catch!, Paul mentioned it in previous review, I missed to update it.